docs/enhancement-proposals/off-ep-0003.md

```diff
@@ -0,0 +1,86 @@
+# OFF-EP 0003 — Allow interpolation of non-integer bond orders
+
+**Status:** Proposed
+
+**Authors:** Matt Thompson, matt.thompson@openforcefield.org
+
+**Stakeholders:** David Mobley, John Chodera, Jeffrey Wagner, Simon Boothroyd
+
+**Acceptance criteria:** Unanimity
+
+**Created:** 2021-11-23
+
+**Discussion:** [Issue #8](https://github.com/openforcefield/standards/issues/8)
+
+**Implementation:** [PR #23](https://github.com/openforcefield/standards/pull/23)
+
+## Abstract
+
+# OFF-EP 0003 — Allow interpolation of non-integer bond orders
+
+This OFF-EP explicitly documents that non-integer bond orders can be used for interpolating parameters.
+
+## Motivation and Scope
+
+The SMIRNOFF specification describes a process by which fractional bond orders can be used to
+interpolate parameters for
+[bond](https://openforcefield.github.io/standards/standards/smirnoff/#fractional-bond-orders) and
+[proper
+torsion](https://openforcefield.github.io/standards/standards/smirnoff/#fractional-torsion-bond-orders)
+forces. The examples provided use bond orders of 1 and 2 and as the basis of interpolation between
+their associated parameters. Work by the Open Force Field Initiative has suggested cases in which it
+would be useful to have non-integer bond orders such that i.e. parameters are interpolated between
+bond orders of 1.0 and 1.2 in a different way than they may be between bond orders of 1.2 and 2.0.
+The current speification does not explicitly forbid non-integer bond orders, but the section
+describes the process of parameter interpolation using an example that does. Also, the current
+implementation in the OpenFF Toolkit does not support non-integer bond orders.
+
+## Usage and Impact
+
+In practice, most bonds in molecules have bond orders very near 1.0 or 2.0 and values such as 1.5
+are much less common. This results in training data being clumped near 1.0 and 2.0 and often the
+slopes of each cluster are not identical. For example, the slope associated with a cluster at 1.0 -
+1.2 might be different from a cluster at 1.8-2.0. @pavankum describes this in more detail in
+[issue #8](https://github.com/openforcefield/standards/issues/8). Being able to have different
+slopes associated with each cluster, i.e. by using 1.0, 1.5, and 2.0 instead of just 1 and 2, would
+enable a force field to account for these differences. Such a parameter may look like
+
+```
+<Bonds version="0.3" potential="harmonic" fractional_bondorder_method="AM1-Wiberg" fractional_bondorder_interpolation="linear">
+    <Bond
+        smirks="[#6X3:1]!#[#6X3:2]"
+        k_bondorder1="820.0*kilocalories_per_mole/angstrom**2"
```

Comment:
Are XML attributes with the `.` valid attribute tokens? Would this hold for other serializations, such as YAML, JSON, BSON, etc? I worry this might cause problems.

If not, it would be much easier to specify interpolation points as
```
k1="820.0*kilocalories_per_mole/angstrom**2" length1="1.45*angstrom" bondorder1="1.0"
k2="1000.0*kilocalories_per_mole/angstrom**2" length2="1.43*angstrom" bondorder2="1.5"
k3="1098*kilocalories_per_mole/angstrom**2" length3="1.35*angstrom" bondorder3="2.0"
```
If we wanted to maintain backward compatibility, we could use
```
k_bondorder1="820.0*kilocalories_per_mole/angstrom**2" length_bondorder1="1.45*angstrom" bondorder1="1.0"
k_bondorder2="1000.0*kilocalories_per_mole/angstrom**2" length_bondorder2="1.43*angstrom" bondorder2="1.5"
k_bondorder3="1098*kilocalories_per_mole/angstrom**2" length_bondorder3="1.35*angstrom" bondorder3="2.0"
```
and have `bondorder1` default to 1.0, `bondorder2` default to 2.0, etc. for backward compatibility and simplicity.

Reply:
Great call. I think that the `bondorderN` solution is exactly what we need here. This sidesteps potential issues with support for `.` characters in different representations. 

> If we wanted to maintain backward compatibility

This may not actually be possible at all. The `Bonds` and `ProperTorsions` section versions will need to be bumped for this update, since this adds a new keyword (`bondorderN`) that wasn't in the old section. (that is, if today's version of the OpenFF toolkit says that it supports `Bonds` section version `0.4`, and the `bondorderN` keyword comes in, then an error will be raised. So the section version must not be `0.4` any more). 

We explicitly forbid reading "serendipitously" valid OFFXML contents that have an unsupported version. This is because the version number can change the default values of optional attributes. So two nearly-identical OFFXML files that differ only by a section version could encode different physics.

But REVERSE compatibility isn't really a problem, because non-time travelers will be more interested in FORWARD compatibility. The current parameter handlers can [recognize and handle a range of old section versions](https://github.com/openforcefield/openff-toolkit/blob/master/openff/toolkit/typing/engines/smirnoff/parameters.py#L2661-L2670) so we should be in good shape there. (that is, `ProperTorsionHandler` can be updated to read every `ProperTorsions` section version from `0.3` to `0.5`)

So, this would lead to a spec bump of `Bonds` and `ProperTorsions` to version `0.5`. I've made an Issue tag to indicate that this will require a section version bump.

> have bondorder1 default to 1.0, bondorder2 default to 2.0, etc

I (begrudgingly) think this is a good idea. It's (begrudging) because implementing optional attributes is always a nightmare, especially when they relate to other optional attributes, and double especially for the forest of indexed, double indexed, and optional attributes that `ProperTorsionHandler` has to process correctly. But I think it's worth it for the user experience, and to have ProperTorsion elements in the OFFXML have a shorter line length.

Reply:
I think the criticisms above are spot-on and the solution here seems like the best way to approach this. I'll work to update the proposal accordingly.

The only bit of context I'd like to add is that we haven't released any force fields with parameter interpolation, so it's possible this could be coordinated to avoid the need to shop an up-converter of existing files from 0.3 or 0.4 to 0.5. There would be a burden on internal users to update, but we that's easier to do than tweak data that's already slipped out to production.